File: Kotlin/src/kotlin/game_of_survival/core/models/factories/CellFactory.kt
package game_of_survival.core.models.factories

import game_of_survival.core.Constants
import game_of_survival.core.abstraction.Cell
import game_of_survival.core.exceptions.ParsingException
import game_of_survival.core.models.cells.HardlyRevivableCell
import game_of_survival.core.models.cells.HibernatingCell
import game_of_survival.core.models.cells.LimitedCell
import game_of_survival.core.models.cells.NormalCell

object CellFactory {
    fun newCell(type: Char, isAlive: Boolean) : Cell = when (type) {
        Constants.CellTypeChars.NORMAL_CELL -> NormalCell(isAlive)
        Constants.CellTypeChars.LIMITED_CELL -> LimitedCell(isAlive)
        Constants.CellTypeChars.HIBERNATING_CELL -> HibernatingCell(isAlive)
        Constants.CellTypeChars.HARDLY_REVIVABLE_CELL -> HardlyRevivableCell(isAlive)
        else -> throw ParsingException("Cell described with `$type` is not supported")
    }
}

File: Cpp/Game_of_survival/cell_factory.cpp
#include "cell_factory.h"

Model::Cell* Model::CellFactory::new_cell(const char type_char, const bool is_alive)
{
	switch (type_char) {
	case Model::NormalCell::TYPE_CHAR:
		return (Model::Cell*)new Model::NormalCell(is_alive);

	case Model::LimitedCell::TYPE_CHAR:
		return (Model::Cell*)new Model::LimitedCell(is_alive);

	case Model::HibernatingCell::TYPE_CHAR:
		return (Model::Cell*)new Model::HibernatingCell(is_alive);

	case Model::HardlyRevivableCell::TYPE_CHAR:
		return (Model::Cell*)new Model::HardlyRevivableCell(is_alive);
	}
	throw Exception::ArgumentException("Wrong type_char");
}
